pallets/farming/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+            withdraw_limit_count: u8,
+        ) -> DispatchResult {
+            T::ControlOrigin::ensure_origin(origin)?;
+
+            let pid = Self::pool_next_id();
+            let keeper = T::Keeper::get().into_sub_account_truncating(pid);
+            let reward_issuer = T::RewardIssuer::get().into_sub_account_truncating(pid);
+            let basic_token = tokens_proportion[0];
+            let tokens_proportion_map: BTreeMap<CurrencyIdOf<T>, Perbill> =
+                tokens_proportion.into_iter().map(|(k, v)| (k, v)).collect();
+            let basic_rewards_map: BTreeMap<CurrencyIdOf<T>, BalanceOf<T>> =
+                basic_rewards.into_iter().map(|(k, v)| (k, v)).collect();
+
+            let mut pool_info = PoolInfo::new(
+                keeper,
+                reward_issuer,
+                tokens_proportion_map,
+                basic_token,
+                basic_rewards_map,
+                None,
+                min_deposit_to_start,
+                after_block_to_start,
+                withdraw_limit_time,
+                claim_limit_time,
+                withdraw_limit_count,
+            );
+
+            if let Some((gauge_token, max_block, gauge_basic_rewards)) = gauge_init {
+                let gauge_basic_rewards_map: BTreeMap<CurrencyIdOf<T>, BalanceOf<T>> =
+                    gauge_basic_rewards
+                        .into_iter()
+                        .map(|(k, v)| (k, v))
+                        .collect();
+
+                Self::create_gauge_pool(
+                    pid,
+                    &mut pool_info,
+                    gauge_token,
+                    gauge_basic_rewards_map,
+                    max_block,
+                )?;
+            };
+
+            PoolInfos::<T>::insert(pid, &pool_info);
+            PoolNextId::<T>::mutate(|id| -> DispatchResult {
+                *id = id.checked_add(1).ok_or(ArithmeticError::Overflow)?;
+                Ok(())
+            })?;
+
+            Self::deposit_event(Event::FarmingPoolCreated { pid });
+            Ok(())
+        }
+
+        /// After create farming pool, need to deposit reward token into the pool.
+        #[pallet::call_index(1)]
+        #[pallet::weight(0)]
+        pub fn charge(
+            origin: OriginFor<T>,
+            pid: PoolId,
+            rewards: Vec<(CurrencyIdOf<T>, BalanceOf<T>)>,
+        ) -> DispatchResult {
+            let exchanger = ensure_signed(origin)?;
+
+            let mut pool_info = Self::pool_infos(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+            rewards
+                .iter()
+                .try_for_each(|(reward_currency, reward)| -> DispatchResult {
+                    T::MultiCurrency::transfer(
+                        *reward_currency,
+                        &exchanger,
+                        &pool_info.reward_issuer,
+                        *reward,
+                    )
+                })?;
+            pool_info.state = PoolState::Charged;
```

Comment:
If user can deposit rewards multiple times, seems it's unnecessary to insert pool_info

Reply:
insert operation here is to change the PoolState.
There's a situation that when pool state is already `Ongoing`, then this charge operation change state to `Charged`, but it's fine, because on next block, it'll go back to `Ongoing` again by `on_initialize`.